ow_dynamic_terrain/src/DynamicTerrainVisual.cpp

```diff
@@ -78,6 +88,53 @@ class DynamicTerrainVisual : public VisualPlugin, public DynamicTerrainBase
   {
     onModifyTerrainMsg(msg, TerrainModifier::modifyPatch);
   }
+
+  void applyNormalMap()
+  {
+    auto heightmap = getHeightmap(get_scene());
+    if (heightmap == nullptr)
+    {
+      gzerr << m_plugin_name << ": DynamicTerrainVisual::applyNormalMap()"
+        " could not acquire heightmap." << endl;
+      return;
+    }
+    auto terrain = heightmap->OgreTerrain()->getTerrain(0, 0);
+
+    // Override custom material normal map labeled "ow_dynamic_terrain_normal_map"
+    // with Ogre::Terrain's internal normal map.
+    // Ogre::Terrain generates an internal normal map and it is updated by
+    // Terrain::updateDerivedData(). But if you use a custom material script
+    // with a custom normal map, the internal normal map will never be seen.
+    // Note: If you are not using a custom material this code will have no
+    // effect, but Ogre should be using the correct normal map and a generated
+    // material and shader in that case.
+    // Caveat: The internal normal map is always sized to the terrain's DEM
+    // resolution, so using a larger texture for more detail is not possible
+    // with this plugin.
+    ResourceManager::ResourceMapIterator res_it = MaterialManager::getSingleton().getResourceIterator();
+    while (res_it.hasMoreElements())
+    {
+      ResourcePtr resource = res_it.getNext();
+      MaterialPtr material = resource.staticCast<Material>();
+      Material::TechniqueIterator tech_it = material->getTechniqueIterator();
+      while (tech_it.hasMoreElements())
+      {
+        Technique* technique = tech_it.getNext();
+        Technique::PassIterator pass_it = technique->getPassIterator();
+        while (pass_it.hasMoreElements())
+        {
+          Pass* pass = pass_it.getNext();
+          TextureUnitState* tus = pass->getTextureUnitState("ow_dynamic_terrain_normal_map");
+          if (tus != 0)
+          {
+            tus->setTexture(terrain->getTerrainNormalMap());
```

Comment:
Is this `setTexture` function capable of being handed a normal map that's larger in resolution than the heightmap, or would that throw and error?

Reply:
Yes, but I think I'm not being clear. If you use a high-res normal map, and then later modify normal maps (which requires a normal map at the same resolution as terrain geometry) then there will be a visual change across the entire workspace as you switch from a high-res map to a lower-res map.